pkg/resourceapply/apps.go

```diff
@@ -17,7 +17,39 @@ func ApplyStatefulSetWithControl(
 	required *appsv1.StatefulSet,
 	options ApplyOptions,
 ) (*appsv1.StatefulSet, bool, error) {
-	return ApplyGeneric[*appsv1.StatefulSet](ctx, control, recorder, required, options)
+	return ApplyGenericWithHandlers[*appsv1.StatefulSet](
+		ctx,
+		control,
+		recorder,
+		required,
+		options,
+		nil,
+		func(required *appsv1.StatefulSet, existing *appsv1.StatefulSet) string {
+			if !equality.Semantic.DeepEqual(existing.Spec.Selector, required.Spec.Selector) {
+				return "spec.selector is immutable"
+			}
+			if !equality.Semantic.DeepEqual(existing.Spec.VolumeClaimTemplates, required.Spec.VolumeClaimTemplates) {
+				return "spec.volumeClaimTemplates is immutable"
+			}
+			if !equality.Semantic.DeepEqual(existing.Spec.ServiceName, required.Spec.ServiceName) {
+				return "spec.serviceName is immutable"
+			}
+			if !equality.Semantic.DeepEqual(existing.Spec.PodManagementPolicy, required.Spec.PodManagementPolicy) {
+				return "spec.podManagementPolicy is immutable"
+			}
+			if !equality.Semantic.DeepEqual(existing.Spec.UpdateStrategy, required.Spec.UpdateStrategy) {
+				return "spec.updateStrategy is immutable"
+			}
+			if !equality.Semantic.DeepEqual(existing.Spec.RevisionHistoryLimit, required.Spec.RevisionHistoryLimit) {
+				return "spec.revisionHistoryLimit is immutable"
+			}
+			if !equality.Semantic.DeepEqual(existing.Spec.Ordinals, required.Spec.Ordinals) {
+				return "spec.ordinals is immutable"
+			}
```

Comment:
I think I'd limit this to explicit immutable changes that we require because of a reason (like when we broke the selector upgrade). In general these are immutable because changing them requires action or has consequences that often needs to be compensated for. Limiting this avoids testing the unused path, debating the procedures and possibly accidentally breaking something in the future. This is more important for sts as it serves the scylla workload.

so in this case: if a selector is still selecting the old pods we can orphan delete the sts and recreate it

Reply:
I agree, done